Add basic coding guidelines

spbnick opened the pull request:
@RH-FMK/developers, I'd like to add a single coding guidelines document for both sktm and skt. Here is something to start with. What do you think? Anything else? What should we add, or perhaps remove? Complying with pycodestyle?

spbnick committed:
Add basic coding guidelines

spbnick commented:
Adding @major, as he wasn't a member before and didn't get my questions above.

veruu commented on CODING.md:
````diff
@@ -0,0 +1,60 @@
+Not this:
+
+```python
+        """
````
This is inconsistent with both PEP 257 (https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) and Google Python guide which mentions it -- the docstring body should start on the line with triple quotes. We should either follow those, or make the difference clear.

spbnick replied:
Yeah, I forgot to mention that. I really don't like how the "standard" way looks. It's also harder to locate the start of the text with your eyes, when it's stuck to these quotes. Do you think we *can* have that exception?

veruu replied:
I also prefer having the docstring body on a separate line when I'm reading the code, but I'm not sure how it plays with automatic documentation tools (which extract text from docstrings). They might generate unexpected newline at the start. So I'd say, if we plan to use those tools check the output first and decide based on that or if we don't, use the version with newline but document it here.

spbnick replied:
I tried both pydoc and Sphinx and the both seem to work fine:

![pydoc_doc](https://user-images.githubusercontent.com/549246/37952899-00cd19b2-31aa-11e8-9dae-de3d2098243f.png)
![sphinx_doc](https://user-images.githubusercontent.com/549246/37952901-00f87382-31aa-11e8-889e-af807efc80c2.png)

In case we have to use a tool which doesn't it should be possible to reformat the text with a regex replace.

veruu replied:
LGTM. Can we have this difference documented in the guideline as well (with reasoning about readability), as long as others agree with this change?

veruu commented:
Definitely mention PEP 8 and pycodestyle as the checker (the pep8 tool which was used previously is not developed anymore and there were some minor changes in the standard which are not reflected there). Also, there is a difference with line length in the Google guide and PEP 8 -- PEP 8 says 79 chars for code (and 72 for comments, but I'm willing to go with 79 for comments as well) while Google guide mentions 80. The difference is not huge but all the checkers will fail because of it.

spbnick commented:
Thanks for the comments and suggestions, @veruu and @major!

I added a mention of PEP8, `pycodestyle`, `pylint` and `flake8`. You didn't mention `pylint`, so I won't be surprised if we don't need it. I'll fix it if that's so. I added a sentence on lengths, and a paragraph on starting docstrings. Anything else we need change/add?

Yes, we need all these checks in CI. We already have `pylint` for tests and `pycodestyle` for everything, and we need to fix the rest of the code and add runs of the remaining tools.

veruu commented:
> You didn't mention pylint, so I won't be surprised if we don't need it. I'll fix it if that's so.

I didn't mention `pylint` because - as you said in the guide - the code is currently not pylint-compliant and it looked unfair to me to add it to the guidelines because of that (additionally, it also complains about things we might not care about (eg a class not having enough methods) so later we might want to ship our own `.pylintrc` which disables some warnings. For now I wouldn't worry about it though, let's keep it the way you wrote the guide).

New changes look good to me and right now I can't think of anything I'd add, but I'll let you know later if I find something.

spbnick commented:
@agustinhenze, do you have anything to add here?

spbnick commented:
Alright, merged. Agustin, please open a PR, if you'd like anything changed here.
